ada/language/ast.py

```diff
@@ -1558,6 +1558,39 @@ def should_ref_generic_formals():
 
     is_in_private_part = Property(Self.parent.parent.is_a(T.PrivatePart))
 
+    @langkit_property(return_type=Bool, public=True)
+    def is_public_decl(origin=T.AdaNode.entity):
+        """
+        Return whether this declaration is public from the point of view of
+        the given ``origin`` node.
+
+        .. ATTENTION::
```

Comment:
It seems to me that to make your `ATTENTION` admonition irrelevant you could have a property more like:

`def is_public(from=T.Unit)` which would avoid your problems of scope visibility because only working at the library level (there would still be a potential misunderstanding about with visibility).

I'm also a bit afraid that this property is too specific, and, frankly, too weird (having to put such an admonition is a bad sign ...). I wonder if a really complete implementation of a real `is_visible_from`, taking into account nesting, with visibililty, etc, wouldn't also allow Guillaume to implement what he needs, while being wildly more useful in the general case. What do you think ?

Reply:
I agree. I updated my PR, which now contains a first implementation of `is_visible`. Although it is not complete, it will report an error if the user tries to call it on an unsupported case instead of returning a wrong answer.

Reply:
Great, thanks Romain!